checks/branch_protected.go

```diff
@@ -31,47 +33,94 @@ func init() {
 	registerCheck(CheckBranchProtection, BranchProtection)
 }
 
+type repositories interface {
+	Get(context.Context, string, string) (*github.Repository,
+		*github.Response, error)
+	ListReleases(ctx context.Context, owner string, repo string, opts *github.ListOptions) (
+		[]*github.RepositoryRelease, *github.Response, error)
+	GetBranchProtection(context.Context, string, string, string) (
+		*github.Protection, *github.Response, error)
+}
+
+type logger func(s string, f ...interface{})
+
 func BranchProtection(c *checker.CheckRequest) checker.CheckResult {
-	repo, _, err := c.Client.Repositories.Get(c.Ctx, c.Owner, c.Repo)
+	// Checks branch protection on both release and development branch
+	return checkReleaseAndDevBranchProtection(c.Ctx, c.Client.Repositories, c.Logf, c.Owner, c.Repo)
+}
+
+func checkReleaseAndDevBranchProtection(ctx context.Context, r repositories, l logger, ownerStr,
+	repoStr string) checker.CheckResult {
+	// Get release branches
+	releases, _, err := r.ListReleases(ctx, ownerStr, repoStr, &github.ListOptions{})
 	if err != nil {
 		return checker.MakeRetryResult(CheckBranchProtection, err)
 	}
 
-	protection, resp, err := c.Client.Repositories.
-		GetBranchProtection(c.Ctx, c.Owner, c.Repo, *repo.DefaultBranch)
+	var checks []checker.CheckResult
+	checkedBranches := map[string]bool{}
+	for _, release := range releases {
+		// may be nil if release tag already exists
+		commitish := release.TargetCommitish
+		if commitish != nil && !checkedBranches[*commitish] {
+			res := getProtectionAndCheck(ctx, r, l, ownerStr, repoStr, *commitish)
+			checkedBranches[*commitish] = true
+			checks = append(checks, res)
+		}
+	}
+
+	// Default development branch check
+	repo, _, err := r.Get(ctx, ownerStr, repoStr)
+	if err != nil {
+		return checker.MakeRetryResult(CheckBranchProtection, err)
+	}
+	if !checkedBranches[*repo.DefaultBranch] {
+		res := getProtectionAndCheck(ctx, r, l, ownerStr, repoStr, *repo.DefaultBranch)
+		checks = append(checks, res)
+	}
+
+	return checker.MultiCheckResultAnd(checks...)
+}
+
+func getProtectionAndCheck(ctx context.Context, r repositories, l logger, ownerStr, repoStr,
```

Comment:
do you know if this will work when the commitish is a commit hash and not a branch name?
For envoy, for example, one of the releases has commitish `"target_commitish": "8fb3cb86082b17144a80402f5367ae65f06083bd"` (see #445)

Reply:
Nice catch! You get a "404 Branch not found" (nil response and an error) as opposed to if there's no protection, which gives "404 Branch not protected" (resp status not found and an error). 

The existing code probably assumed the default branch was always there. Updating so this skips when branches are not found.

Reply:
I think you could check if a commitish is a hash. If it is, retrieve the name of the branch the commits belongs to, and use the branch name to call `getProtectionAndCheck()`.

Reply:
I refactored to check whether the branch exists with ListBranches. If it doesn't exist/is deleted, it skips.

I'm not sure I can check which branch the commit belongs to. I found references to the github search api, which seems like it's in preview mode.

Reply:
> I refactored to check whether the branch exists with ListBranches. If it doesn't exist/is deleted, it skips.

shall we fail when the commitish points to a non-existing branch, since it provides no information about the protection of the branch that was used? Wdut?

> I'm not sure I can check which branch the commit belongs to. I found references to the github search api, which seems like it's in preview mode.

you're right, does not seem to work. An alternative would be to clone the repo and use `git branch --all --contains <commit>`. I think it's better to wait for the API above to be more broadly available. Let's check if the commitish looks like a hash and ignore this case for now then. Let's create a github issue to track this for future improvement.

Any better ideas?

Reply:
> > I refactored to check whether the branch exists with ListBranches. If it doesn't exist/is deleted, it skips.
> 
> shall we fail when the commitish points to a non-existing branch, since it provides no information about the protection of the branch that was used? Wdut?

while looking at envoy, I realized you renamed the `master` branch to `main` (the new naming that GitHub recommends). With what I suggested above, we would fail the test since we would not find the master branch no longer exists. If you go with the solution I suggested, you'll have to treat main and master as being the same branch. 

> > I'm not sure I can check which branch the commit belongs to. I found references to the github search api, which seems like it's in preview mode.
> 
> you're right, does not seem to work. An alternative would be to clone the repo and use `git branch --all --contains <commit>`. I think it's better to wait for the API above to be more broadly available. Let's check if the commitish looks like a hash and ignore this case for now then. Let's create a github issue to track this for future improvement.
> 
> Any better ideas?

Reply:
Sorry for the delay, was doing a bunch of local testing and ran into some rate limits. To answer the questions:
* Agreed, let's ignore the hash commitish for now (I added a comment TODO to go back later for improvement)
* About the branch rename. Ugh! the Github REST API gives you a 301 moved permanently when you try to get a moved branch (like master). BUT! Go-github does NOT from my testing. It gives you a 404 Not Found, giving me no way to distinguish between a moved and a deleted/non-existing branch. It feels wrong to fail on that. Still trying to see if I can get moved information from go-github.

Reply:
> Sorry for the delay, was doing a bunch of local testing and ran into some rate limits. To answer the questions:
> 
> * Agreed, let's ignore the hash commitish for now (I added a comment TODO to go back later for improvement)
+1

> * About the branch rename. Ugh! the Github REST API gives you a 301 moved permanently when you try to get a moved branch (like master). BUT! Go-github does NOT from my testing. It gives you a 404 Not Found, giving me no way to distinguish between a moved and a deleted/non-existing branch. It feels wrong to fail on that. Still trying to see if I can get moved information from go-github.

nice finding. I did not know about that! While you investigate, we could file an github issue for go-github to ask about the behavior and whether it'd be a breaking change for them to fix it. 
Last resort would be for us to make the http request ourselves :/ Kinda of sad to have to do this, though

Reply:
Yep, just did! https://github.com/google/go-github/issues/1895

For now just to keep this PR going I'll manually handle the master->main change, and otherwise we'll just fail when the commitish returns a non-existing branch.

Reply:
done... as much as can do.